src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp

```diff
@@ -0,0 +1,666 @@
+                                                       strainIncrement,
+                                                       stack.totalStress,
+                                                       stack.dTotalStress_dPressure,
+                                                       stack.dTotalStress_dTemperature,
+                                                       stack.stiffness,
+                                                       porosity,
+                                                       porosity_n,
+                                                       dPorosity_dVolStrain,
+                                                       dPorosity_dPressure,
+                                                       dPorosity_dTemperature,
+                                                       dSolidDensity_dPressure );
+
+  // Step 2: compute the body force
+  if( m_gravityAcceleration > 0.0 )
+  {
+    computeBodyForce( k, q,
+                      porosity,
+                      dPorosity_dVolStrain,
+                      dPorosity_dPressure,
+                      dPorosity_dTemperature,
+                      dSolidDensity_dPressure,
+                      stack );
+  }
+
+  // Step 3: compute fluid mass increment
+  computeFluidIncrement( k, q,
+                         porosity,
+                         porosity_n,
+                         dPorosity_dVolStrain,
+                         dPorosity_dPressure,
+                         dPorosity_dTemperature,
+                         stack );
+
+  // Step 4: compute pore volume constraint
+  computePoreVolumeConstraint( k,
+                               porosity,
+                               dPorosity_dVolStrain,
+                               dPorosity_dPressure,
+                               dPorosity_dTemperature,
+                               stack );
+}
+
+
+template< typename SUBREGION_TYPE,
+          typename CONSTITUTIVE_TYPE,
+          typename FE_TYPE >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void ThermalMultiphasePoromechanics< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >::
+computeBodyForce( localIndex const k,
+                  localIndex const q,
+                  real64 const & porosity,
+                  real64 const & dPorosity_dVolStrain,
+                  real64 const & dPorosity_dPressure,
+                  real64 const & dPorosity_dTemperature,
+                  real64 const & dSolidDensity_dPressure,
+                  StackVariables & stack ) const
+{
+  using Deriv = constitutive::multifluid::DerivativeOffset;
+
+  Base::computeBodyForce( k, q,
+                          porosity,
+                          dPorosity_dVolStrain,
+                          dPorosity_dPressure,
+                          dPorosity_dTemperature,
+                          dSolidDensity_dPressure,
+                          stack, [&]( real64 const & totalMassDensity,
+                                      real64 const & mixtureDensity )
```

Comment:
what is the difference between totalMassDensity and mixtureDensity?

Reply:
mixtureDensity that is the average between fluid and solid density is commonly named by bulk density

Reply:
Yes